File: minigit-5/ts/src/objects.ts
// Git object storage: Blob, Tree, Commit, Tag handling

import * as fs from 'fs';
import * as path from 'path';
import { sha1, compress, decompress, getMinigitDir, ensureDir, modeToString } from './utils';
import { BlobObject, TreeObject, CommitObject, TagObject, GitObject, TreeEntry } from './types';

export function hashObject(type: string, content: Buffer): string {
  const header = Buffer.from(`${type} ${content.length}\0`);
  const store = Buffer.concat([header, content]);
  return sha1(store);
}

export function writeObject(type: string, content: Buffer, repoRoot?: string): string {
  const header = Buffer.from(`${type} ${content.length}\0`);
  const store = Buffer.concat([header, content]);
  const objectSha = sha1(store);

  const minigitDir = getMinigitDir(repoRoot);
  const objectDir = path.join(minigitDir, 'objects', objectSha.slice(0, 2));
  ensureDir(objectDir);

  const objectPath = path.join(objectDir, objectSha.slice(2));
  if (!fs.existsSync(objectPath)) {
    const compressed = compress(store);
    fs.writeFileSync(objectPath, compressed);
  }

  return objectSha;
}

export function readObject(sha: string, repoRoot?: string): { type: string; content: Buffer } {
  const minigitDir = getMinigitDir(repoRoot);
  const objectPath = path.join(minigitDir, 'objects', sha.slice(0, 2), sha.slice(2));

  if (!fs.existsSync(objectPath)) {
    throw new Error(`Object not found: ${sha}`);
  }

  const compressed = fs.readFileSync(objectPath);
  const raw = decompress(compressed);

  // Parse header: "type size\0content"
  const nullIndex = raw.indexOf(0);
  if (nullIndex === -1) {
    throw new Error(`Invalid object: ${sha}`);
  }

  const header = raw.slice(0, nullIndex).toString();
  const [type] = header.split(' ');
  const content = raw.slice(nullIndex + 1);

  return { type, content };
}

export function objectExists(sha: string, repoRoot?: string): boolean {
  try {
    const minigitDir = getMinigitDir(repoRoot);
    const objectPath = path.join(minigitDir, 'objects', sha.slice(0, 2), sha.slice(2));
    return fs.existsSync(objectPath);
  } catch {
    return false;
  }
}

export function expandShortSha(shortSha: string, repoRoot?: string): string {
  if (shortSha.length === 40) {
    return shortSha;
  }

  const minigitDir = getMinigitDir(repoRoot);
  const objectsDir = path.join(minigitDir, 'objects');

  if (shortSha.length < 4) {
    throw new Error(`SHA too short: ${shortSha}`);
  }

  const prefix = shortSha.slice(0, 2);
  const rest = shortSha.slice(2);
  const searchDir = path.join(objectsDir, prefix);

  if (!fs.existsSync(searchDir)) {
    throw new Error(`Object not found: ${shortSha}`);
  }

  const matches = fs.readdirSync(searchDir).filter(name => name.startsWith(rest));

  if (matches.length === 0) {
    throw new Error(`Object not found: ${shortSha}`);
  }

  if (matches.length > 1) {
    throw new Error(`Ambiguous SHA: ${shortSha}`);
  }

  return prefix + matches[0];
}

export function createBlob(content: Buffer, write: boolean = false, repoRoot?: string): string {
  if (write) {
    return writeObject('blob', content, repoRoot);
  }
  return hashObject('blob', content);
}

export function parseTree(content: Buffer): TreeEntry[] {
  const entries: TreeEntry[] = [];
  let offset = 0;

  while (offset < content.length) {
    // Find space after mode
    const spaceIndex = content.indexOf(0x20, offset);
    if (spaceIndex === -1) break;

    const mode = content.slice(offset, spaceIndex).toString();

    // Find null after name
    const nullIndex = content.indexOf(0, spaceIndex + 1);
    if (nullIndex === -1) break;

    const name = content.slice(spaceIndex + 1, nullIndex).toString();

    // Next 20 bytes are SHA (binary)
    const shaBytes = content.slice(nullIndex + 1, nullIndex + 21);
    const sha = shaBytes.toString('hex');

    entries.push({ mode, name, sha });
    offset = nullIndex + 21;
  }

  return entries;
}

export function serializeTree(entries: TreeEntry[]): Buffer {
  // Sort entries: directories (trees) come after files with same prefix
  const sorted = entries.slice().sort((a, b) => {
    // For sorting, append / to directory names (mode 40000)
    const aName = a.mode === '40000' ? a.name + '/' : a.name;
    const bName = b.mode === '40000' ? b.name + '/' : b.name;
    return aName.localeCompare(bName);
  });

  const parts: Buffer[] = [];
  for (const entry of sorted) {
    // Mode (no leading zeros for directories: 40000 not 040000)
    const mode = entry.mode === '40000' ? '40000' : entry.mode;
    const header = Buffer.from(`${mode} ${entry.name}\0`);
    const sha = Buffer.from(entry.sha, 'hex');
    parts.push(header, sha);
  }

  return Buffer.concat(parts);
}

export function createTree(entries: TreeEntry[], repoRoot?: string): string {
  const content = serializeTree(entries);
  return writeObject('tree', content, repoRoot);
}

export function parseCommit(content: Buffer): CommitObject {
  const lines = content.toString().split('\n');
  let tree = '';
  const parents: string[] = [];
  let author = '';
  let committer = '';
  let message = '';
  let inMessage = false;

  for (let i = 0; i < lines.length; i++) {
    const line = lines[i];

    if (inMessage) {
      message += (message ? '\n' : '') + line;
    } else if (line === '') {
      inMessage = true;
    } else if (line.startsWith('tree ')) {
      tree = line.slice(5);
    } else if (line.startsWith('parent ')) {
      parents.push(line.slice(7));
    } else if (line.startsWith('author ')) {
      author = line.slice(7);
    } else if (line.startsWith('committer ')) {
      committer = line.slice(10);
    }
  }

  return { type: 'commit', tree, parents, author, committer, message };
}

export function serializeCommit(commit: Omit<CommitObject, 'type'>): Buffer {
  let content = `tree ${commit.tree}\n`;
  for (const parent of commit.parents) {
    content += `parent ${parent}\n`;
  }
  content += `author ${commit.author}\n`;
  content += `committer ${commit.committer}\n`;
  content += `\n${commit.message}`;

  return Buffer.from(content);
}

export function createCommit(
  tree: string,
  parents: string[],
  author: string,
  committer: string,
  message: string,
  repoRoot?: string
): string {
  const content = serializeCommit({ tree, parents, author, committer, message });
  return writeObject('commit', content, repoRoot);
}

export function parseTag(content: Buffer): TagObject {
  const lines = content.toString().split('\n');
  let object = '';
  let objectType = '';
  let tagName = '';
  let tagger = '';
  let message = '';
  let inMessage = false;

  for (let i = 0; i < lines.length; i++) {
    const line = lines[i];

    if (inMessage) {
      message += (message ? '\n' : '') + line;
    } else if (line === '') {
      inMessage = true;
    } else if (line.startsWith('object ')) {
      object = line.slice(7);
    } else if (line.startsWith('type ')) {
      objectType = line.slice(5);
    } else if (line.startsWith('tag ')) {
      tagName = line.slice(4);
    } else if (line.startsWith('tagger ')) {
      tagger = line.slice(7);
    }
  }

  return { type: 'tag', object, objectType, tagName, tagger, message };
}

export function serializeTag(tag: Omit<TagObject, 'type'>): Buffer {
  let content = `object ${tag.object}\n`;
  content += `type ${tag.objectType}\n`;
  content += `tag ${tag.tagName}\n`;
  content += `tagger ${tag.tagger}\n`;
  content += `\n${tag.message}`;

  return Buffer.from(content);
}

export function createTag(
  object: string,
  objectType: string,
  tagName: string,
  tagger: string,
  message: string,
  repoRoot?: string
): string {
  const content = serializeTag({ object, objectType, tagName, tagger, message });
  return writeObject('tag', content, repoRoot);
}

export function getObjectType(sha: string, repoRoot?: string): string {
  const { type } = readObject(sha, repoRoot);
  return type;
}

export function getObjectSize(sha: string, repoRoot?: string): number {
  const { content } = readObject(sha, repoRoot);
  return content.length;
}

export function getBlob(sha: string, repoRoot?: string): Buffer {
  const { type, content } = readObject(sha, repoRoot);
  if (type !== 'blob') {
    throw new Error(`Expected blob, got ${type}`);
  }
  return content;
}

export function getTree(sha: string, repoRoot?: string): TreeEntry[] {
  const { type, content } = readObject(sha, repoRoot);
  if (type !== 'tree') {
    throw new Error(`Expected tree, got ${type}`);
  }
  return parseTree(content);
}

export function getCommit(sha: string, repoRoot?: string): CommitObject {
  const { type, content } = readObject(sha, repoRoot);
  if (type !== 'commit') {
    throw new Error(`Expected commit, got ${type}`);
  }
  return parseCommit(content);
}

export function getTagObject(sha: string, repoRoot?: string): TagObject {
  const { type, content } = readObject(sha, repoRoot);
  if (type !== 'tag') {
    throw new Error(`Expected tag, got ${type}`);
  }
  return parseTag(content);
}

// Get tree SHA from a tree-ish (commit, tag, or tree)
export function getTreeFromTreeIsh(sha: string, repoRoot?: string): string {
  const { type, content } = readObject(sha, repoRoot);

  if (type === 'tree') {
    return sha;
  }

  if (type === 'commit') {
    const commit = parseCommit(content);
    return commit.tree;
  }

  if (type === 'tag') {
    const tag = parseTag(content);
    return getTreeFromTreeIsh(tag.object, repoRoot);
  }

  throw new Error(`Cannot get tree from ${type}`);
}

// Walk a tree recursively and return all blob paths with their SHAs
export function walkTree(
  treeSha: string,
  prefix: string = '',
  repoRoot?: string
): Map<string, { sha: string; mode: string }> {
  const result = new Map<string, { sha: string; mode: string }>();
  const entries = getTree(treeSha, repoRoot);

  for (const entry of entries) {
    const fullPath = prefix ? `${prefix}/${entry.name}` : entry.name;

    if (entry.mode === '40000') {
      // Directory - recurse
      const subMap = walkTree(entry.sha, fullPath, repoRoot);
      subMap.forEach((value, key) => result.set(key, value));
    } else {
      result.set(fullPath, { sha: entry.sha, mode: entry.mode });
    }
  }

  return result;
}

File: minigit-5/py/objects.py
"""Git object handling: Blob, Tree, Commit, Tag."""

from pathlib import Path
from typing import Optional, List, Tuple
import os

from utils import (
    write_object, read_object, hash_object_data,
    git_dir, resolve_short_sha
)


class Blob:
    """Represents a git blob (file content)."""

    def __init__(self, data: bytes):
        self.data = data
        self._sha: Optional[str] = None

    @property
    def sha(self) -> str:
        if self._sha is None:
            self._sha = hash_object_data('blob', self.data)
        return self._sha

    def write(self, repo_root: Optional[Path] = None) -> str:
        self._sha = write_object('blob', self.data, repo_root)
        return self._sha

    @classmethod
    def read(cls, sha: str, repo_root: Optional[Path] = None) -> 'Blob':
        obj_type, data = read_object(sha, repo_root)
        if obj_type != 'blob':
            raise ValueError(f"Object {sha} is not a blob")
        blob = cls(data)
        blob._sha = sha
        return blob


class TreeEntry:
    """A single entry in a tree object."""

    def __init__(self, mode: int, name: str, sha: str):
        self.mode = mode
        self.name = name
        self.sha = sha

    @property
    def type(self) -> str:
        if self.mode == 0o40000:
            return 'tree'
        else:
            return 'blob'


class Tree:
    """Represents a git tree (directory)."""

    def __init__(self, entries: Optional[List[TreeEntry]] = None):
        self.entries = sorted(entries or [], key=lambda e: e.name)
        self._sha: Optional[str] = None

    def serialize(self) -> bytes:
        """Serialize tree to bytes."""
        result = b''
        for entry in sorted(self.entries, key=lambda e: e.name + ('/' if e.mode == 0o40000 else '')):
            mode_str = f"{entry.mode:o}"
            sha_bytes = bytes.fromhex(entry.sha)
            result += f"{mode_str} {entry.name}\0".encode() + sha_bytes
        return result

    @property
    def sha(self) -> str:
        if self._sha is None:
            self._sha = hash_object_data('tree', self.serialize())
        return self._sha

    def write(self, repo_root: Optional[Path] = None) -> str:
        self._sha = write_object('tree', self.serialize(), repo_root)
        return self._sha

    @classmethod
    def read(cls, sha: str, repo_root: Optional[Path] = None) -> 'Tree':
        obj_type, data = read_object(sha, repo_root)
        if obj_type != 'tree':
            raise ValueError(f"Object {sha} is not a tree")

        entries = []
        i = 0
        while i < len(data):
            # Find space after mode
            space_idx = data.index(b' ', i)
            mode = int(data[i:space_idx].decode(), 8)

            # Find null after name
            null_idx = data.index(b'\0', space_idx)
            name = data[space_idx + 1:null_idx].decode()

            # Next 20 bytes are SHA
            sha_bytes = data[null_idx + 1:null_idx + 21]
            entry_sha = sha_bytes.hex()

            entries.append(TreeEntry(mode, name, entry_sha))
            i = null_idx + 21

        tree = cls(entries)
        tree._sha = sha
        return tree


class Commit:
    """Represents a git commit."""

    def __init__(
        self,
        tree_sha: str,
        parents: List[str],
        author: str,
        author_email: str,
        author_date: str,
        committer: str,
        committer_email: str,
        committer_date: str,
        message: str
    ):
        self.tree_sha = tree_sha
        self.parents = parents
        self.author = author
        self.author_email = author_email
        self.author_date = author_date
        self.committer = committer
        self.committer_email = committer_email
        self.committer_date = committer_date
        self.message = message
        self._sha: Optional[str] = None

    def serialize(self) -> bytes:
        """Serialize commit to bytes."""
        lines = []
        lines.append(f"tree {self.tree_sha}")
        for parent in self.parents:
            lines.append(f"parent {parent}")
        lines.append(f"author {self.author} <{self.author_email}> {self.author_date}")
        lines.append(f"committer {self.committer} <{self.committer_email}> {self.committer_date}")
        lines.append("")
        lines.append(self.message)

        return '\n'.join(lines).encode()

    @property
    def sha(self) -> str:
        if self._sha is None:
            self._sha = hash_object_data('commit', self.serialize())
        return self._sha

    def write(self, repo_root: Optional[Path] = None) -> str:
        self._sha = write_object('commit', self.serialize(), repo_root)
        return self._sha

    @classmethod
    def read(cls, sha: str, repo_root: Optional[Path] = None) -> 'Commit':
        obj_type, data = read_object(sha, repo_root)
        if obj_type != 'commit':
            raise ValueError(f"Object {sha} is not a commit")

        text = data.decode()
        lines = text.split('\n')

        tree_sha = None
        parents = []
        author = author_email = author_date = None
        committer = committer_email = committer_date = None
        message_start = 0

        for i, line in enumerate(lines):
            if line == '':
                message_start = i + 1
                break
            elif line.startswith('tree '):
                tree_sha = line[5:]
            elif line.startswith('parent '):
                parents.append(line[7:])
            elif line.startswith('author '):
                author_info = line[7:]
                author, rest = parse_author_line(author_info)
                author_email = rest[0]
                author_date = rest[1]
            elif line.startswith('committer '):
                committer_info = line[10:]
                committer, rest = parse_author_line(committer_info)
                committer_email = rest[0]
                committer_date = rest[1]

        message = '\n'.join(lines[message_start:])

        commit = cls(
            tree_sha, parents,
            author, author_email, author_date,
            committer, committer_email, committer_date,
            message
        )
        commit._sha = sha
        return commit


def parse_author_line(line: str) -> Tuple[str, Tuple[str, str]]:
    """Parse author/committer line: 'Name <email> timestamp tz'"""
    # Find the < for email start
    lt_idx = line.index('<')
    gt_idx = line.index('>')

    name = line[:lt_idx].strip()
    email = line[lt_idx + 1:gt_idx]
    date_part = line[gt_idx + 1:].strip()

    return name, (email, date_part)


class Tag:
    """Represents an annotated git tag."""

    def __init__(
        self,
        object_sha: str,
        object_type: str,
        tag_name: str,
        tagger: str,
        tagger_email: str,
        tagger_date: str,
        message: str
    ):
        self.object_sha = object_sha
        self.object_type = object_type
        self.tag_name = tag_name
        self.tagger = tagger
        self.tagger_email = tagger_email
        self.tagger_date = tagger_date
        self.message = message
        self._sha: Optional[str] = None

    def serialize(self) -> bytes:
        """Serialize tag to bytes."""
        lines = []
        lines.append(f"object {self.object_sha}")
        lines.append(f"type {self.object_type}")
        lines.append(f"tag {self.tag_name}")
        lines.append(f"tagger {self.tagger} <{self.tagger_email}> {self.tagger_date}")
        lines.append("")
        lines.append(self.message)

        return '\n'.join(lines).encode()

    @property
    def sha(self) -> str:
        if self._sha is None:
            self._sha = hash_object_data('tag', self.serialize())
        return self._sha

    def write(self, repo_root: Optional[Path] = None) -> str:
        self._sha = write_object('tag', self.serialize(), repo_root)
        return self._sha

    @classmethod
    def read(cls, sha: str, repo_root: Optional[Path] = None) -> 'Tag':
        obj_type, data = read_object(sha, repo_root)
        if obj_type != 'tag':
            raise ValueError(f"Object {sha} is not a tag")

        text = data.decode()
        lines = text.split('\n')

        object_sha = None
        tag_object_type = None
        tag_name = None
        tagger = tagger_email = tagger_date = None
        message_start = 0

        for i, line in enumerate(lines):
            if line == '':
                message_start = i + 1
                break
            elif line.startswith('object '):
                object_sha = line[7:]
            elif line.startswith('type '):
                tag_object_type = line[5:]
            elif line.startswith('tag '):
                tag_name = line[4:]
            elif line.startswith('tagger '):
                tagger_info = line[7:]
                tagger, rest = parse_author_line(tagger_info)
                tagger_email = rest[0]
                tagger_date = rest[1]

        message = '\n'.join(lines[message_start:])

        tag = cls(
            object_sha, tag_object_type, tag_name,
            tagger, tagger_email, tagger_date,
            message
        )
        tag._sha = sha
        return tag
